### PR TITLE
Experiment with array creation.

### PR DESCRIPTION
**Pull request by CyrusNajmabadi**

### CODE REVIEW

File: src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/Utilities/FixedSizeArrayBuilder.cs [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections.Immutable;
+using System.Runtime.InteropServices;
+using Microsoft.CodeAnalysis.PooledObjects;
+using Roslyn.Utilities;
+
+/// <summary>
+/// A bare-bones, pooled builder, focused on the case of producing <see cref="ImmutableArray{T}"/>s where the final
```
Comment (ToddGrun) [inline src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/Utilities/FixedSizeArrayBuilder.cs:12]: >pooled builder [](http://example.com/codeflow?start=18&length=14)

?

File: src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/Utilities/FixedSizeArrayBuilder.cs [hunk 2]
```diff
@@ -0,0 +1,42 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections.Immutable;
+using System.Runtime.InteropServices;
+using Microsoft.CodeAnalysis.PooledObjects;
```
Comment (ToddGrun) [inline src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/Utilities/FixedSizeArrayBuilder.cs:8]: needed?

File: src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/Utilities/FixedSizeArrayBuilder.cs [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+    /// is an error for a client of this type to specify a capacity and then attempt to call <see
+    /// cref="MoveToImmutable"/> without that number of elements actually having been added to the builder.  This will
+    /// throw if attempted.  This <see cref="FixedSizeArrayBuilder{T}"/> is effectively unusable once this is called.
+    /// The internal buffer will reset to an empty array, meaning no more items could ever be added to it.
+    /// </summary>
+    public ImmutableArray<T> MoveToImmutable()
+    {
+        Contract.ThrowIfTrue(_index != _values.Length);
+        var result = ImmutableCollectionsMarshal.AsImmutableArray(_values);
+        _values = Array.Empty<T>();
```
Comment (ToddGrun) [inline src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/Utilities/FixedSizeArrayBuilder.cs:38]: >Array.Empty<T>(); [](http://example.com/codeflow?start=18&length=17)

was there a reason you didn't want to use a collection initializer here?

**Review by ToddGrun — APPROVED** [review]
:shipit:

File: src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/Utilities/FixedSizeArrayBuilder.cs (hunk 1 above)
Comment (CyrusNajmabadi) [reply src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/Utilities/FixedSizeArrayBuilder.cs:12]: will fix that in followup pr.

File: src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/Utilities/FixedSizeArrayBuilder.cs (hunk 3 above)
Comment (CyrusNajmabadi) [reply src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/Utilities/FixedSizeArrayBuilder.cs:38]: yes.  i liked the clarity of "you are getting this exact instance".

File: src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/Utilities/FixedSizeArrayBuilder.cs (hunk 2 above)
Comment (CyrusNajmabadi) [reply src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/Utilities/FixedSizeArrayBuilder.cs:8]: nope.  removed in followup.

**Pull request merged**